httpx/concurrency/asyncio.py

```diff
@@ -60,58 +59,27 @@ def get_http_version(self) -> str:
         ident = ssl_object.selected_alpn_protocol()
         return "HTTP/2" if ident == "h2" else "HTTP/1.1"
 
-    async def read(
-        self, n: int, timeout: TimeoutConfig = None, flag: TimeoutFlag = None
-    ) -> bytes:
-        if timeout is None:
-            timeout = self.timeout
-
-        while True:
-            # Check our flag at the first possible moment, and use a fine
-            # grained retry loop if we're not yet in read-timeout mode.
-            should_raise = flag is None or flag.raise_on_read_timeout
-            read_timeout = timeout.read_timeout if should_raise else 0.01
-            try:
-                data = await asyncio.wait_for(self.stream_reader.read(n), read_timeout)
-                break
-            except asyncio.TimeoutError:
-                if should_raise:
-                    raise ReadTimeout() from None
-                # FIX(py3.6): yield control back to the event loop to give it a chance
-                # to cancel `.read(n)` before we retry.
-                # This prevents concurrent `.read()` calls, which asyncio
-                # doesn't seem to allow on 3.6.
-                # See: https://github.com/encode/httpx/issues/382
-                await asyncio.sleep(0)
-
-        return data
-
-    def write_no_block(self, data: bytes) -> None:
-        self.stream_writer.write(data)  # pragma: nocover
-
-    async def write(
-        self, data: bytes, timeout: TimeoutConfig = None, flag: TimeoutFlag = None
-    ) -> None:
-        if not data:
-            return
+    async def read_or_timeout(self, n: int, timeout: typing.Optional[float]) -> bytes:
+        try:
+            return await asyncio.wait_for(self.stream_reader.read(n), timeout)
+        except asyncio.TimeoutError:
+            # FIX(py3.6): yield control back to the event loop to give it a chance
+            # to cancel `.read(n)` before we fail and possibly retry.
+            # This prevents concurrent `.read()` calls, which asyncio
+            # doesn't seem to allow on 3.6.
+            # See: https://github.com/encode/httpx/issues/382
+            await asyncio.sleep(0)
 
-        if timeout is None:
-            timeout = self.timeout
+            raise ReadTimeout() from None
 
+    async def write_or_timeout(
+        self, data: bytes, timeout: typing.Optional[float]
+    ) -> None:
         self.stream_writer.write(data)
-        while True:
```

Comment:
This is subtly different now and I'm not sure I understand how looping on the write is safe. Previously the `while True` retry loop encapsulated just the `drain()` call, but with this change both the `write` and `drain` are inside the loop.

The asyncio docs say of the `write()` method:
> The method attempts to write the data to the underlying socket immediately. If that fails, the data is queued in an internal write buffer until it can be sent.

If we retry on write how can we be sure we aren't writing duplicated data to that internal write buffer?